aggregator/src/aggregator.rs

```diff
@@ -3459,6 +3633,222 @@ impl VdafOps {
+        max_batch_size: usize,
+        max_batch_write_delay: StdDuration,
+    ) {
+        let mut is_done = false;
+        let mut batch_expiry = TokioInstant::now();
+        let mut report_writers = Vec::with_capacity(max_batch_size);
+        let mut result_txs = Vec::with_capacity(max_batch_size);
+        while !is_done {
+            // Wait for an event of interest.
+            let write_batch = select! {
+                // Wait until we receive a report to be written (or the channel is closed due to the
+                // ReportWriteBatcher being dropped)...
+                item = report_rx.recv() => {
+                    match item {
+                        // We got an item. Add it to the current batch of reports to be written.
+                        Some((report_writer, rslt_tx)) => {
+                            if report_writers.is_empty() {
+                                batch_expiry = TokioInstant::now() + max_batch_write_delay;
+                            }
+                            report_writers.push(report_writer);
+                            result_txs.push(rslt_tx);
+                            report_writers.len() >= max_batch_size
+                        }
+
+                        // The channel is closed. Note this, and write any final reports that may be
+                        // batched before shutting down.
+                        None => {
+                            is_done = true;
+                            !report_writers.is_empty()
+                        },
+                    }
+                },
+
+                // ... or the current batch times out.
+                _ = sleep_until(batch_expiry), if !report_writers.is_empty() => true,
+            };
+
+            // If the event made us want to write the current batch to storage, do so.
+            if write_batch {
+                let ds = Arc::clone(&ds);
+                let result_writers =
+                    replace(&mut report_writers, Vec::with_capacity(max_batch_size));
+                let result_txs = replace(&mut result_txs, Vec::with_capacity(max_batch_size));
+                tokio::spawn(async move {
+                    Self::write_batch(ds, result_writers, result_txs).await;
+                });
+            }
+        }
+    }
+
+    async fn write_batch(
+        ds: Arc<Datastore<C>>,
+        report_writers: Vec<Box<dyn ReportWriter<C>>>,
+        result_txs: Vec<oneshot::Sender<Result<(), Arc<Error>>>>,
+    ) {
+        // Check preconditions.
+        assert_eq!(report_writers.len(), result_txs.len());
+
+        // Run all report writes concurrently.
+        let report_writers = Arc::new(report_writers);
+        let rslts = ds
+            .run_tx_with_name("upload", |tx| {
+                let report_writers = Arc::clone(&report_writers);
+                Box::pin(async move {
+                    Ok(join_all(report_writers.iter().map(|rw| rw.write_report(tx))).await)
```

Comment:
This outermost Result is always Ok, so only the top branch of the below match block gets taken. Moreover, `run_tx()` cannot see past this outermost result, so it would not be able to retry on serialization failures. In other places, we use `gather_errors(join_all(...))`, but I don't think we want that here, because we don't want one `Error::MutationTargetAlreadyExisted` to rollback the entire transaction. We'll need a new function that selectively hoists serialization failures and deadlock failures from a vector-of-results to the top level of a result-of-vector-of-results.

Reply:
The outer `Result` will be `Err` if the overall commit fails (and only then, I believe). In this case, we would return the same error to all waiting uploaders.

That's a good point about serialization failures. However, the suggested solution is kind of messy: even if we had the suggested function, we'd end up writing the logic to use it in the `ReportBatchWriter`.

I think a cleaner solution might be to change how our serialization-failure detection works. We currently assume that if a serialization failure is returned from some Postgres query function, it will be returned from the callback passed to `run_tx`. But, of course, this isn't necessarily true -- and we've been trying to paper over it via solutions like our custom `try_join` macro, `gather_errors`, etc. Instead, perhaps we can try a different solution: everywhere we get an error back from the database, check it at that point & set a "retry" flag on the `datastore::Transaction`; if this flag is set, we trigger a retry as we currently would based on seeing a serialization failure from the callback. This allows the callback to do whatever error handling it desires, while still getting the expected retry-on-serialization-failure behavior. (I suppose this would keep the callback from purposefully _avoiding_ a retry by returning a different error, but I don't think we need to do that anyway; and we could implement this feature a different way, if desirable.)

WDYT? If you agree, unless you object I might do this in a follow-on PR as it is only somewhat related to this PR & is likely to be a fairly large change.

Reply:
(er, to summarize: the big advantage of my suggested solution, as I see it, is that it keeps the complexity of tracking retries inside of the `Datastore` & `Transaction` logic, rather than pushing it out to folks using the datastore to run transactions. I think it's somewhat unmaintainable to expect every callback to intelligently filter multiple errors up to the one that the `Datastore`/`Transaction` code would _like_ to see.)

Reply:
Tracking errors solely from inside the datastore module sounds like a good strategy going forward. While I won't insist on it, it might be nice to at least add a for loop to propagate serialization failures in this PR, so that we can more easily measure the performance impacts of the various improvements here.

Reply:
Fair enough -- I've added some hacky serialization-failure checking logic to `ReportBatchWriter`. I'll rip it out with the next PR, but it should be good enough for performance testing.